front/lib/api/assistant/conversation.ts

```diff
@@ -6,20 +6,314 @@ import {
+      email: userMessage.userContextEmail,
+      profilePictureUrl: userMessage.userContextProfilePictureUrl,
+    },
+  };
+}
+
+async function renderAgentMessage(
+  auth: Authenticator,
+  message: Message,
+  agentMessage: AgentMessage
+): Promise<AgentMessageType> {
+  const [agentConfiguration, agentRetrievalAction] = await Promise.all([
+    AgentConfiguration.findOne({
+      where: {
+        id: agentMessage.agentConfigurationId,
+      },
+    }),
+    (async () => {
+      if (agentMessage.agentRetrievalActionId) {
+        return await renderRetrievalActionByModelId(
+          agentMessage.agentRetrievalActionId
+        );
+      }
+      return null;
+    })(),
+  ]);
+
+  if (!agentConfiguration) {
+    throw new Error(
+      `Agent configuration ${agentMessage.agentConfigurationId} not found`
+    );
+  }
+
+  const configuration = await getAgentConfiguration(
+    auth,
+    agentConfiguration.sId
+  );
+
+  return {
+    id: message.id,
+    sId: message.sId,
+    type: "agent_message",
+    visibility: message.visibility,
+    version: message.version,
+    parentMessageId: null,
+    status: agentMessage.status,
+    action: agentRetrievalAction,
+    message: agentMessage.message,
+    feedbacks: [],
+    error: null,
+    configuration,
+  };
+}
+
+export async function getConversation(
+  auth: Authenticator,
+  conversationId: string
+): Promise<ConversationType | null> {
+  const conversation = await Conversation.findOne({
+    where: {
+      sId: conversationId,
+    },
+  });
+
+  if (!conversation) {
+    return null;
+  }
+
+  const messages = await Message.findAll({
+    where: {
+      conversationId: conversation.id,
+    },
+    order: [
+      ["rank", "ASC"],
+      ["version", "ASC"],
+    ],
+    include: [
+      {
+        model: UserMessage,
+        as: "userMessage",
+        required: false,
+      },
+      {
+        model: AgentMessage,
+        as: "agentMessage",
+        required: false,
+      },
+    ],
+  });
+
+  const render = await Promise.all(
+    messages.map((message) => {
+      return (async () => {
+        if (message.userMessage) {
+          const m = await renderUserMessage(auth, message, message.userMessage);
+          return { m, rank: message.rank, version: message.version };
+        }
+        if (message.agentMessage) {
+          const m = await renderAgentMessage(
+            auth,
+            message,
+            message.agentMessage
+          );
+          return { m, rank: message.rank, version: message.version };
+        }
+        throw new Error("Unreachable: message must be either user or agent");
+      })();
+    })
+  );
+
+  render.sort((a, b) => {
+    if (a.rank !== b.rank) {
+      return a.rank - b.rank;
+    }
+    return a.version - b.version;
+  });
+
+  // We need to escape the type system here to create content.
+  const content: any[] = Array.from(
+    { length: messages.reduce((acc, m) => Math.max(acc, m.rank), -1) + 1 },
```

Comment:
would explain in a comment why we need the max rank rather than just message.length (also helps to understand the line is about getting the array size, not creating content)

Reply:
comment added above